hip-0000-smart-contract-hcs-interoperable.md

```diff
@@ -0,0 +1,58 @@
+---
+hip: 478
+title: Interoperability Between Smart Contracts and the HCS
+author: John Conway (@scalemaildev), Walter Hernandez (@walter-hernandez)
+type: Standards Track
+category: Core
+needs-council-approval: Yes
+status: Last Call
+last-call-date-time: 2022-06-21T07:00:00Z
+created: 05-17-2022
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/479
+updated: 05-24-2022
+---
+
+## Abstract
+
+Enable smart contract interaction with the Hedera Consensus Service, by allowing contracts to create, read, and write to HCS topics.
+
+## Motivation
+
+Interoperability between the HCS and smart contracts would enable developers to leverage HCS-stored data in their dApps.
+
+## Rationale
+
+1. By enabling smart contracts to interact with the HCS, this would allow the creation of HCS-based Oracles. For example, a smart contract could subscribe to a private or public topic in order to read its external data, such as the prices of crypto-assets.
```

Comment:
Can you elaborate on the topic subscription.
Currently that is offered through a gRPC service on the mirror node only.
How would you envision it be supported at the smart contract layer?

Reply:
@Nana-EC My thinking is that it could be:

- A topic subscription could be a combination of scheduled transactions with Topics. For example, a transaction could be scheduled to read the latest message in a topic and send the message to a smart contract. The scheduled transactions are the middleman to carry the message from the topic to the function in the smart contract

Of course, the limitations above is:

- I think Hedera's Scheduled Transactions may not have interoperability with Hedera Smart Contract Service
- In the scenario of Hedera's Scheduled Transactions being possible to integrate with Hedera Smart Contract Service, there could be some difficulties for signing of messages sent from Hedera's Scheduled Transactions to  Hedera Smart Contract Service

The approach of using gRPC at the mirror node works quite well for topic subscription and integration with a solution outside of Hedera Smart Contract Service. 

Let me know what you and if you know if there is a better way to do it or if my thinking/assumption may be flawed

Reply:
@walter-hernandez thanks for the details. 
Similar to a comment before I believe you'll want to expand this HIP with the details (descriptions and diagrams) of the precompile contracts and the network to highlight capability.
Especially as it comes to the scheduled component. 
One thing to note is, the network does not store topic messages in state, only the topic information highlighted in [Consensus Topic Info](https://github.com/hashgraph/hedera-protobufs/blob/main/services/consensus_topic_info.proto).

The message queueing service probably needs its own separate discussion.

Reply:
HCS is not strictly like a queue management system such as Kafka, it doesn't inherently know who is subscribing to it and doesn't know if any subscriber did get the message, it's up to the subscriber to keep track of the last message timestamp they received and query a mirror node from that timestamp onwards.

Providing an automated data feed into HSCS directly from HCS bring up issues of the TPS differential between HCS and Smart Contracts (10k vs 350k).

You could consider that a specific topic could be throttled to match the HSCS speed, but that would entail Hedera knowing which topics are subscribed to by smart contracts, knowing that a message on a topic will result in x gas consumption, etc... which would not be trivial (I suspect impossible).

Added to which, executing a smart contract requires gas which isn't specified in the topic message submission.

The best model as far as I can tell is for an external oracle to subscribe to HCS messages and deal with making the contract calls.